OADP-4058: Adding structure for OADP and 3scale user story (#85994)
openshift/openshift-docs: modules/restoring-the-mysql-database.adoc

:_mod-docs-content-type: PROCEDURE

//included in backing-up-and-restoring-3scale-using-oadp.adoc assembly

[id="restoring-the-mysql-database_{context}"]
= Restoring the mysql database

The restoration of the mysql database re-creates the Pod, ReplicationController, DeploymentConfig, additional PV, and persistent volume claim (PVC).

NOTE: Do not delete the default PV and PVC associated with the database. 

.Prerequisites

* You have restored the secrets and APIManager custom resources.

.Procedure

. Scale down the threescale operator:
+
[source,terminal]
----
$ oc scale deployment threescale-operator-controller-manager-v2 --replicas=0 -n threescale
deployment.apps/threescale-operator-controller-manager-v2 scaled
----

> . Scale down all the DeploymentConfigs:

Reviewer: ```suggestion
. Scale down all the `DeploymentConfig` objects by creating and running the following script:
```

> +

Reviewer: .. Create the script by using the following example:

[source,terminal]
----
$ cat ./scaledowndc.sh

for dc in apicast-production apicast-staging backend-cron backend-listener backend-redis backend-worker system-app system-memcache system-mysql system-redis system-searchd system-sidekiq zync zync-database zync-que; do
    oc scale dc/$dc --replicas=0 -n threescale
done
----

> +

Reviewer: ```suggestion
.. Run the script you created by using the following command:
```
Or something like this. It is not one step and one command; we're clearly doing two things here.

Reviewer 2: Agree with "Run the script  you created". But we should drop "by using the following command" :-)

[source,terminal]
----

> $ ./scaledowndc.sh 

Reviewer: ```suggestion
$ ./scaledowndc.sh 
```
Hmm, is this a valid command?

> 

Reviewer: ```suggestion
+
.Example output
[source,terminal]
----
```

deploymentconfig.apps.openshift.io/apicast-production scaled
deploymentconfig.apps.openshift.io/apicast-staging scaled
deploymentconfig.apps.openshift.io/backend-cron scaled
deploymentconfig.apps.openshift.io/backend-listener scaled
deploymentconfig.apps.openshift.io/backend-redis scaled
deploymentconfig.apps.openshift.io/backend-worker scaled
deploymentconfig.apps.openshift.io/system-app scaled
deploymentconfig.apps.openshift.io/system-memcache scaled
deploymentconfig.apps.openshift.io/system-mysql scaled
deploymentconfig.apps.openshift.io/system-redis scaled
deploymentconfig.apps.openshift.io/system-searchd scaled
deploymentconfig.apps.openshift.io/system-sidekiq scaled
deploymentconfig.apps.openshift.io/zync scaled
deploymentconfig.apps.openshift.io/zync-database scaled
deploymentconfig.apps.openshift.io/zync-que scaled
----

> . Delete the system-mysql DeploymentConfig:

Reviewer: ```suggestion
. Delete the `system-mysql` `DeploymentConfig` object by running the following command:
```

+
[source,terminal]
----
$ oc delete dc system-mysql -n threescale

> 

Reviewer: again, separate input from output

Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
deploymentconfig.apps.openshift.io "system-mysql" deleted
deploymentconfig.apps.openshift.io "backend-redis" deleted
----

> . Restore the mysql database using YAML:

Reviewer: ```suggestion
. Restore the `mysql` database with a YAML file by using the following command:
```

+
[source,terminal]
----
$ oc create -f restore-backend.yaml
----
+
[source,yaml]
----
apiVersion: velero.io/v1
kind: Restore
metadata:

>   name: restore-mysql

Reviewer: Should the `name` here match the preceding command, or the `backupName` below?

  namespace: openshift-adp
spec:
  backupName: mysql-backup
  excludedResources:
    - nodes
    - events
    - events.events.k8s.io
    - backups.velero.io
    - restores.velero.io
    - csinodes.storage.k8s.io
    - volumeattachments.storage.k8s.io
    - backuprepositories.velero.io
    - resticrepositories.velero.io
  hooks:
    resources:
      - name: restoreDB
        postHooks:
          - exec:
              command:
                - /bin/sh
                - '-c'
                - >
                  sleep 30
                  
                  mysql -h 127.0.0.1 -D system -u root
                  --password=$MYSQL_ROOT_PASSWORD <
                  /var/lib/mysqldump/data/dump.sql
              container: system-mysql
              execTimeout: 80s
              onError: Fail
              waitTimeout: 5m
  itemOperationTimeout: 1h0m0s
  restorePVs: true
----

.Verification 

> . Verify if the `PodVolumeRestore` is completed:

Reviewer: ```suggestion
. Verify that the `PodVolumeRestore` is completed by running the following command:
```

+
[source,terminal]
----
$ oc get podvolumerestore -n openshift-adp

> 

Reviewer: separate input v output

NAME                    NAMESPACE    POD                     UPLOADER TYPE   VOLUME                  STATUS      TOTALBYTES   BYTESDONE   AGE

restore-mysql-rbzvm     threescale   system-mysql-2-kjkhl    kopia           mysql-storage           Completed   771879108    771879108   40m
restore-mysql-z7x7l     threescale   system-mysql-2-kjkhl    kopia           example-claim           Completed   380415       380415      40m
----

> . Verify the additional persistent volume claim (PVC) has also been restored:

Reviewer: ```suggestion
. Verify that the additional persistent volume claim (PVC) has also been restored by running the following command:
```

+
[source,terminal]
----
$ oc get pvc -n threescale

> 

Reviewer: separate input v output

NAME                    STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   VOLUMEATTRIBUTESCLASS   AGE
backend-redis-storage   Bound    pvc-3dca410d-3b9f-49d4-aebf-75f47152e09d   1Gi        RWO            gp3-csi        <unset>                 68m
example-claim           Bound    pvc-cbaa49b0-06cd-4b1a-9e90-0ef755c67a54   1Gi        RWO            gp3-csi        <unset>                 57m
mysql-storage           Bound    pvc-4549649f-b9ad-44f7-8f67-dd6b9dbb3896   1Gi        RWO            gp3-csi        <unset>                 68m
system-redis-storage    Bound    pvc-04dadafd-8a3e-4d00-8381-6041800a24fc   1Gi        RWO            gp3-csi        <unset>                 68m
system-searchd          Bound    pvc-afbf606c-d4a8-4041-8ec6-54c5baf1a3b9   1Gi        RWO            gp3-csi        <unset>                 68m
----

> .Next steps

Reviewer: ```suggestion
.Next step
```

:_mod-docs-content-type: PROCEDURE

//included in backing-up-and-restoring-3scale-using-oadp.adoc assembly

[id="restoring-the-mysql-database_{context}"]
= Restoring the mysql database

> The restoration of the `mysql` database recreates the `Pod`, `ReplicationController`, and deployment objects, plus the additional persistent volumes (PVs) and associated persistent volume claims (PVCs). 
> It restores the `mysql` dump, which the `example-claim` persistent volume claim (PVC) contains.

Reviewer 2: ```suggestion
Restoration of the `mysql` database causes the following processes:

* Recreation of the `Pod`, `ReplicationController`, and deployment objects.
* Recreation of the additional persistent volumes (PVs) and associated persistent volume claims (PVCs). 
* Restoration of the `mysql` dump, which the `example-claim` persistent volume claim (PVC) contains.
```
A suggestion for better readability. Feel free to rephrase even more ;-)

Author: Thanks, that's an awesome suggestion :)

WARNING: Do not delete the default PV and PVC associated with the database. If you delete the volumes or volume claims, your backups are deleted.

.Prerequisites

* You restored the `Secrets` and `APIManager` custom resources (CR).

.Procedure

. Scale down the threescale operator:
+
[source,terminal]
----
$ oc scale deployment threescale-operator-controller-manager-v2 --replicas=0 -n threescale
----
+
.Example output:
[source,terminal]
----
deployment.apps/threescale-operator-controller-manager-v2 scaled
----

. Create the following script to scale down the 3scale operator:
+
[source,terminal]
----
$ cat ./scaledowndeployment.sh
----
+
.Example output:
[source,terminal]
----
for dc in apicast-production apicast-staging backend-cron backend-listener backend-redis backend-worker system-app system-memcache system-mysql system-redis system-searchd system-sidekiq zync zync-database zync-que; do
    oc scale dc/$dc --replicas=0 -n threescale
done
----

. Scale down all the deployment 3scale components by running the following script:
+
[source,terminal]
----
$ ./scaledowndeployment.sh 
----
+
.Example output:
[source,terminal]
----

> deploymentconfig.apps.openshift.io/apicast-production scaled

Author: @stillalearner , could you please confirm if it should be deployment instead of "deploymentconfig" in the output?
Thanks!

deploymentconfig.apps.openshift.io/apicast-staging scaled
deploymentconfig.apps.openshift.io/backend-cron scaled
deploymentconfig.apps.openshift.io/backend-listener scaled
deploymentconfig.apps.openshift.io/backend-redis scaled
deploymentconfig.apps.openshift.io/backend-worker scaled
deploymentconfig.apps.openshift.io/system-app scaled
deploymentconfig.apps.openshift.io/system-memcache scaled
deploymentconfig.apps.openshift.io/system-mysql scaled
deploymentconfig.apps.openshift.io/system-redis scaled
deploymentconfig.apps.openshift.io/system-searchd scaled
deploymentconfig.apps.openshift.io/system-sidekiq scaled
deploymentconfig.apps.openshift.io/zync scaled
deploymentconfig.apps.openshift.io/zync-database scaled
deploymentconfig.apps.openshift.io/zync-que scaled
----

> . Delete the `system-mysql` deployment object by running the following command:

Reviewer 2: ```suggestion
. Delete the `system-mysql` deployment object:
```

deploymentconfig.apps.openshift.io "system-mysql" deleted
----

. Create the following YAML file to restore the `mysql` database:
+
.Example restore-mysql.yaml file
[source,yaml]
----
apiVersion: velero.io/v1
kind: Restore
metadata:
  name: restore-mysql
  namespace: openshift-adp
spec:
  backupName: mysql-backup
  excludedResources:
    - nodes
    - events
    - events.events.k8s.io
    - backups.velero.io
    - restores.velero.io
    - csinodes.storage.k8s.io
    - volumeattachments.storage.k8s.io
    - backuprepositories.velero.io
    - resticrepositories.velero.io
  hooks:
    resources:
      - name: restoreDB
        postHooks:
          - exec:
              command:
                - /bin/sh
                - '-c'
                - >
                  sleep 30
                  
                  mysql -h 127.0.0.1 -D system -u root
                  --password=$MYSQL_ROOT_PASSWORD <
                  /var/lib/mysqldump/data/dump.sql <1>
              container: system-mysql
              execTimeout: 80s
              onError: Fail
              waitTimeout: 5m
  itemOperationTimeout: 1h0m0s
  restorePVs: true
----
<1> The data is restored from this path.

> . Restore the `mysql` database with by running the following command:

Reviewer 2: ```suggestion
. Restore the `mysql` database:
```

+
[source,terminal]
----
$ oc create -f <restore-mysql>.yaml
----

.Verification 

> . Verify that the `PodVolumeRestore` is completed by running the following command:

Reviewer 2: ```suggestion
. Verify that the `PodVolumeRestore` is completed:
```

+
[source,terminal]
----
$ oc get podvolumerestore -n openshift-adp
----
+
.Example output:
[source,terminal]
----
NAME                    NAMESPACE    POD                     UPLOADER TYPE   VOLUME                  STATUS      TOTALBYTES   BYTESDONE   AGE

restore-mysql-rbzvm     threescale   system-mysql-2-kjkhl    kopia           mysql-storage           Completed   771879108    771879108   40m
restore-mysql-z7x7l     threescale   system-mysql-2-kjkhl    kopia           example-claim           Completed   380415       380415      40m
----

> . Verify that the additional persistent volume claim (PVC) has also been restored by running the following command:

Reviewer 2: ```suggestion
. Verify that the additional persistent volume claim (PVC) has been restored:
```

+
[source,terminal]
----
$ oc get pvc -n threescale
----
+
.Example output:
[source,terminal]
----
NAME                    STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   VOLUMEATTRIBUTESCLASS   AGE
backend-redis-storage   Bound    pvc-3dca410d-3b9f-49d4-aebf-75f47152e09d   1Gi        RWO            gp3-csi        <unset>                 68m
example-claim           Bound    pvc-cbaa49b0-06cd-4b1a-9e90-0ef755c67a54   1Gi        RWO            gp3-csi        <unset>                 57m
mysql-storage           Bound    pvc-4549649f-b9ad-44f7-8f67-dd6b9dbb3896   1Gi        RWO            gp3-csi        <unset>                 68m
system-redis-storage    Bound    pvc-04dadafd-8a3e-4d00-8381-6041800a24fc   1Gi        RWO            gp3-csi        <unset>                 68m
system-searchd          Bound    pvc-afbf606c-d4a8-4041-8ec6-54c5baf1a3b9   1Gi        RWO            gp3-csi        <unset>                 68m
----

.Next step

> * Restore the backend Redis database.

Reviewer 2: ```suggestion
* Restore the back-end Redis database.
```
IBM SG for the adjective

:_mod-docs-content-type: PROCEDURE

//included in backing-up-and-restoring-3scale-by-using-oadp.adoc assembly

[id="restoring-the-mysql-database_{context}"]
= Restoring the mysql database

Restoration of the `mysql` database causes the following processes:

* Recreation of the `Pod`, `ReplicationController`, and deployment objects.
* Recreation of the additional persistent volumes (PVs) and associated persistent volume claims (PVCs). 
* Recreation of the `mysql` dump, which the `example-claim` persistent volume claim (PVC) contains.

WARNING: Do not delete the default PV and PVC associated with the database. If you delete the volumes or volume claims, your backups are deleted.

.Prerequisites

* You restored the `Secrets` and `APIManager` custom resources (CR).

.Procedure

. Scale down the threescale operator:
+
[source,terminal]
----
$ oc scale deployment threescale-operator-controller-manager-v2 --replicas=0 -n threescale
----
+
.Example output:
[source,terminal]
----
deployment.apps/threescale-operator-controller-manager-v2 scaled
----

. Create the following script to scale down the 3scale operator:
+
[source,terminal]
----

> $ cat ./scaledowndeployment.sh

Author: I've fixed the command.
